engine/classes/ElggMenuBuilder.php

```diff
@@ -108,35 +108,53 @@ protected function setupTrees() {
 		foreach ($this->menu as $key => $section) {
 			$parents = array();
 			$children = array();
+			$all_menu_items = array();
+			
 			// divide base nodes from children
 			foreach ($section as $menu_item) {
 				/* @var \ElggMenuItem $menu_item */
 				$parent_name = $menu_item->getParentName();
+				$menu_item_name = $menu_item->getName();
+				
 				if (!$parent_name) {
-					$parents[$menu_item->getName()] = $menu_item;
+					// no parents so top level menu items
+					$parents[$menu_item_name] = $menu_item;
 				} else {
-					$children[] = $menu_item;
+					$children[$menu_item_name] = $menu_item;
 				}
+				
+				$all_menu_items[$menu_item_name] = $menu_item;
+			} 
+			
+			if (empty($all_menu_items)) {
+				// empty sections can be skipped
+				continue;
 			}
-
-			// attach children to parents
-			$iteration = 0;
-			$current_gen = $parents;
-			$next_gen = null;
-			while (count($children) && $iteration < 20) {
-				foreach ($children as $index => $menu_item) {
-					$parent_name = $menu_item->getParentName();
-					if (array_key_exists($parent_name, $current_gen)) {
-						$next_gen[$menu_item->getName()] = $menu_item;
-						if (!in_array($menu_item, $current_gen[$parent_name]->getData('children'))) {
-							$current_gen[$parent_name]->addChild($menu_item);
-							$menu_item->setParent($current_gen[$parent_name]);
-						}
-						unset($children[$index]);
-					}
+						
+			if (empty($parents)) {
+				// menu items without parents? That is sad.. report to the log
```

Comment:
top level menu items go into $parents, so you can't have top-level and empty parents, but only sublevel menu items without parents... that is what this notice is for.